src/Shell.cpp

```diff
@@ -0,0 +1,366 @@
+        std::string inputFile, outputFile;
+        parseCommand(commands[i], args, inputFile, outputFile);
+
+        pid_t pid = fork();
+        if (pid == 0) {
+            if (i > 0) {
+                if (dup2(pipefds[j-2], 0) < 0) {  // Redirect stdin to the previous pipe's read end
+                    perror("dup2");
+                    exit(EXIT_FAILURE);
+                }
+            }
+
+            if (i < numCommands - 1) {
+                if (dup2(pipefds[j+1], 1) < 0) {  // Redirect stdout to the next pipe's write end
+                    perror("dup2");
+                    exit(EXIT_FAILURE);
+                }
+            }
+
+            for (int k = 0; k < 2 * (numCommands - 1); k++) {
+                close(pipefds[k]);
+            }
+
+            handleRedirection(inputFile, outputFile);
+
+            std::vector<char*> c_args;
+            for (const auto& arg : args) {
+                c_args.push_back(const_cast<char*>(arg.c_str()));
+            }
+            c_args.push_back(nullptr);
+
+            std::string fullPath;
+            if (!Utils::findExe(c_args[0], fullPath)) {
+                std::cerr << "Command not found: " << c_args[0] << std::endl;
+                exit(EXIT_FAILURE);
+            }
+
+            execv(fullPath.c_str(), c_args.data());
+            perror("execv");
+            exit(EXIT_FAILURE);
+        } else if (pid < 0) {
+            perror("fork");
+            exit(EXIT_FAILURE);
+        }
+        
+        if (i == numCommands - 1 && isBackground) {
+            addJob(pid, commands[i]);
+        }
+
+        j += 2;
+    }
+
+    for (int i = 0; i < 2 * (numCommands - 1); i++) {
+        close(pipefds[i]);
+    }
+
+    if (!isBackground) {
+        for (int i = 0; i < numCommands; i++) {
+            wait(NULL);
+        }
+    }
+}
+
+// this function takes two strings as input and redirects the input and output of the shell to the files
+void Shell::handleRedirection(const std::string &inputFile, const std::string &outputFile)
+{
+    if (!inputFile.empty()) 
+    {
+        int fd = open(inputFile.c_str(), O_RDONLY); 
+        if (fd == -1) 
+        {
+            perror("open input file failed");
+            exit(-1);
+        }
+        dup2(fd, STDIN_FILENO);  // Redirect STDIN
+        close(fd);
+    }
+
+    if (!outputFile.empty()) 
+    {
+        int fd = open(outputFile.c_str(), O_WRONLY | O_CREAT | O_TRUNC, 0644); //readonly , truncates the file to 0 , 0644 is the permission for the file
+        if (fd == -1) 
+        {
+            perror("open output file failed");
+            exit(-1);
+        }
+        dup2(fd, STDOUT_FILENO);  // Redirect STDOUT
+        close(fd);
+    }
+}
+
+// my main method for this shell program
+// it goes over the input string and processes it accordingly
+// it checks for in program commands and also path commands
+void Shell::inputLoop() {
+    bool running = true;
+    while (running) {
```

Comment:
It's still `while (true)` which is not a good practice. hint: you could use the `exit`.